### ISSUE TITLE
Templated Controls Tutorial?

### ISSUE THREAD

**Issue by TheColonel2688** [AvaloniaUI/Avalonia#2516]
Is there a Tutorial for creating Templated Controls? If not can this be added to the to-do?

**Comment by Gillibald** [comment]
Good idea 👍

**Comment by grokys** [comment]
Yes, we definitely need this! However it's very similar to the WPF way of doing it except for how we define properties and that's described here: http://avaloniaui.net/docs/authoring-controls/defining-properties so in the meantime you might be able to muddle through!

**Comment by JulianKowalczuk** [comment]
This is what I used to create control with nested children and some additional properties. I'm using ContentControl, but it inherits from TemplatedControl.
`
<ContentControl xmlns="https://github.com/avaloniaui" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:d="http://schemas.microsoft.com/expression/blend/2008" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" x:Class="MyNamespace.Controls.PinkLabelPanel" xmlns:controls="clr-namespace:MyNamespace.Controls">
  <ContentControl.Styles>
    <Style Selector="controls|PinkLabelPanel">
      <Setter Property="Template">
        <ControlTemplate>
          <Panel>
            <Border Padding="5">
              <StackPanel Background="#180316">
                <Border Padding="0,10,0,0">
                  <StackPanel HorizontalAlignment="Center">
                    <Image Source="{TemplateBinding Label}" Height="40"/>
                    <Image Source="/assets/panel/headerPinkUnderlinePanel.png" Height="10" Width="{TemplateBinding LabelUnderlineWidth}" Stretch="Fill"/>
                  </StackPanel>
                </Border>
                <Border Padding="15">
                  <StackPanel Orientation="Horizontal" Spacing="30" HorizontalAlignment="Center" VerticalAlignment="Center">
                    <StackPanel HorizontalAlignment="Center" Width="200">
                        <ContentPresenter Content="{TemplateBinding Content}"/>
                    </StackPanel>
                  </StackPanel>
                </Border>
              </StackPanel>
            </Border>
            <Image Source="/assets/panel/leftCornerPanel.png" HorizontalAlignment="Left" VerticalAlignment="Top" Height="46"/>
            <Image Source="/assets/panel/rightCornerPanel.png" HorizontalAlignment="Right" VerticalAlignment="Top" Height="46"/>
          </Panel>
        </ControlTemplate>
      </Setter>
    </Style>
  </ContentControl.Styles>
</ContentControl>
`
`
using Avalonia;
using Avalonia.Controls;
using Avalonia.Markup.Xaml;
using Avalonia.Media.Imaging;

namespace HackshieldLauncher.Controls
{
  public class PinkLabelPanel : ContentControl
  {
    public static readonly StyledProperty<IBitmap> LabelProperty =
        AvaloniaProperty.Register<PinkLabelPanel, IBitmap>(nameof(Label));

    public IBitmap Label
    {
      get { return GetValue(LabelProperty); }
      set { SetValue(LabelProperty, value); }
    }

    public static readonly StyledProperty<int> LabelUnderlineWidthProperty =
    AvaloniaProperty.Register<PinkLabelPanel, int>(nameof(LabelUnderlineWidth));

    public int LabelUnderlineWidth
    {
      get { return GetValue(LabelUnderlineWidthProperty); }
      set { SetValue(LabelUnderlineWidthProperty, value); }
    }

    public PinkLabelPanel()
    {
      this.InitializeComponent();
    }

    private void InitializeComponent()
    {
      AvaloniaXamlLoader.Load(this);
    }
  }
}
`

and usage example:
`
<controls:PinkLabelPanel Label="/assets/screens/myProfile/informationPinkHeader.png" LabelUnderlineWidth="270">
      <TextBlock Classes="h2">KURWA MAMY TO</TextBlock>
    </controls:PinkLabelPanel>
`

**Comment by maxkatz6** [comment]
We now have templated control template.

**Comment by tanveerbadar** [comment]
@JulianKowalczuk You are a life-saver, building upon your example and customizing to my needs worked like a charm.

**Comment by lekrieg** [comment]
Here is an actual functional example:

**My control**
`<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:Playground.Controls">
    <Design.PreviewWith>
        <controls:MyControl />
    </Design.PreviewWith>

    <Style Selector="controls|MyControl">
        
        <Setter Property="Template">
            <ControlTemplate>
                <Border BorderThickness="2" BorderBrush="Blue">
                    <ContentPresenter Content="{TemplateBinding Content}"/>
                </Border>
            </ControlTemplate>
        </Setter>
    </Style>
</Styles>
`

**My control code behind**
`using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.Primitives;

namespace Playground.Controls;

public class MyControl : ContentControl
{
}`

**How we can prepare the control for use in the app.axaml**
`<Application.Styles>
        <FluentTheme />
        <StyleInclude Source="Controls/MyControl.axaml"/>
    </Application.Styles>`
    
    **An actual window using it**
    `<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:Playground.ViewModels"
        xmlns:controls="using:Playground.Controls"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="Playground.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel"
        Icon="/Assets/avalonia-logo.ico"
        Title="Playground">

    <Design.DataContext>
        
        <vm:MainWindowViewModel/>
    </Design.DataContext>

    <StackPanel>
        <controls:MyControl Width="100">
            <StackPanel>
                <TextBlock Text="Its working!"/>
                <TextBlock Text="Lol"/>
            </StackPanel>
        </controls:MyControl>
        <TextBlock Text="teasd"/>
    </StackPanel>
    
</Window>
`

Basically you just need to inherit from ContentControl, which has a Content property, that can receive a content, no matter what it is.

**Comment by lekrieg** [comment]
![image](https://github.com/AvaloniaUI/Avalonia/assets/23697245/ef6fbfbf-a4c5-45aa-b7ce-20ea8f09d2d9)

**Comment by remi-scarlet-1** [comment]
This should not be closed.
It is still heavily underdocumented...

**Comment by timunie** [comment]
@remi-scarlet-1 

1. please also check: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CustomControls/RatingControlSample 
2. Feel free to open a docs request on Avalonia-docs repo with details what you are missing. Also PRs to the docs are welcome once you understood what the missing part was.

**Comment by remi-scarlet-1** [comment]
I cant find any documentation about having custom xml elements with content, like how the default element ListBox does ListBoxItem. I'm new to the framework and xml in general so I don't know if I'm missing out on anything.

I have tried creating CustomUser controls but I quickly noticed they weren't designed for that purpose.
Then I tried using "lookless" templated controls which did work, but not with logical children, rather only with a source attribute like: <cc:MyCustomControl ItemSource:"Item1,Item2,Item3"/>

I also tried to find the xml implementation of ListBox in the Avalonia main repo but couldn't.

Is there any documentation of this specific use of logical descendant xml elements?

**Comment by lekrieg** [comment]
> I cant find any documentation about having custom xml elements with content, like how the default element ListBox does ListBoxItem. I'm new to the framework and xml in general so I don't know if I'm missing out on anything.
> 
> ```
> <ListBox>
>     <ListBoxItem>Test1</ListBoxItem>
>     <ListBoxItem>Test2</ListBoxItem>
>     <ListBoxItem>Test3</ListBoxItem>
> </ListBox>
> ```
> 
> I have tried:
> 
>     * creating CustomUser controls but I quickly noticed they weren't designed for that purpose.
> 
>     * using "lookless" templated controls which did work, but not with logical children, rather only with a source attribute: `<cc:MyCustomControl ItemSource:"Item1,Item2,Item3"/>`
> 
> 
> I also tried to find the xml implementation of ListBox in the Avalonia main repo but couldn't. Is there any documentation of this specific use of logical descendant xml elements?

Well, you have 2 choices if you want a styled ListBox.
1 - you can u use a ControlTheme and modify the items to your preferences
2 - Create a component using a TemplatedControl, hinherit from ListBox / ListItem or a control and do it in the same way


one uses DictionaryResource and modify the visual / behavior and the other offers you more control.

I can write about it in HashNode if you want, I am doing some crazy stuff at work these days.

**Comment by lekrieg** [comment]
You can have a look on Semi-Avalonia too, they have some very nice examples about how to modify some stuff

**Comment by Gorvaine** [comment]
I don't know if this is the right place to ask this question. If it's better somewhere else, please let me know.

As a test, I want to achieve a TextBox background color change when the word "Dog" is written on the textbox. I can do it with more properties and by implementing the logic in the parent's ViewModel, but I was trying to do it inside the TemplatedControl. I'm not sure if it's even possible 😅

I'm trying to get an event when text is written on a TextBox which is inside a TemplatedControl to then, modify the background color. But I'm not finding that event.

I've got the StyledProperty:
`    public static readonly StyledProperty<string> Capt2Property =
        AvaloniaProperty.Register<GenControl, string>(nameof(Capt2), defaultValue: "", defaultBindingMode: Avalonia.Data.BindingMode.TwoWay);

    public string Capt2
    {
      get => GetValue(Capt2Property);
      set => SetValue(Capt2Property, value);
    }`

Bound with 
`<TextBox FontSize="16" Text="{TemplateBinding Capt2}" />`

I tried attaching to `OnPropertyChanged`. It triggers when the mouse enters inside the TextBox and other events, but I can not find the event modifying my Capt2Property when I type anything on the textbox.

Also, I tried created a ViewModel for the TemplatedControl with a Capt2 [ObservableProperty] and it worked, I just had to change my links from:
`<TextBox FontSize="16" Text="{TemplateBinding Capt2}" />`
to
`<TextBox FontSize="16" Text="{Binding Capt2}" />`

But then I lost the capability of using the property on the TemplatedControl parent.
`        <ctlrs:GenControl
            Grid.Row="1"
            Capt2="Home" />`

Thanks.

**Comment by Gorvaine** [comment]
Monday sweet, monday.

Just changed the property from a StyledProperty to a DirectProperty. Ensure the mode=TwoWay.
`  public static readonly DirectProperty<GenControl, string> Capt2Property =
      AvaloniaProperty.RegisterDirect<GenControl, string>(
        nameof(Capt2),
        o => o.Capt2,
        (o, v) => { o.Capt2 = v; ChangeBackground(v); },
        defaultBindingMode: BindingMode.TwoWay);

  private string _capt2;
  public string Capt2
  {
    get { return _capt2; }
    set { SetAndRaise(Capt2Property, ref _capt2, value); }
  }`

And now I'm getting the event on the "onPropertyChanged" and on its own DirectProperty Setter.

Sorry for the notifications. I think we have better the documentation for this aspect on:
https://github.com/AvaloniaUI/Avalonia.Samples/blob/main/src/Avalonia.Samples/CustomControls/RatingControlSample/README.adoc

than the official:
https://docs.avaloniaui.net/docs/basics/user-interface/controls/creating-controls/defining-properties

**Comment by lekrieg** [comment]
Did you try to use a multibinding converter or a validator / property validate to check and modify the text color? Maybe you could have more control on your templated control in this way.

**Comment by rpc-scandinavia** [comment]
I wanted to create button with a icon before the text, and after a lot of trial and error I ended up with the code below.
I thought that I could just "inherit" Button, but it was not that easy.

There is a `Button` inside my `IconButton` control, so I map all properties and events common to `Button` and `ContentControl`.

But I want my `IconButton` to look like a button with a buttons initial property values, so I need to copy the property values. But because the `OnApplyTemplate` eventhandler is fired after the properties has been set to the `IconButton`, I need to remember which properties has been set, and only copy the unchanged property values from `Button` to my `IconButton`.

Perhaps there exist a better way, but for now this work for me.

AXAML in "IconButton.axaml"

```
<ContentControl
	xmlns="https://github.com/avaloniaui"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	xmlns:ui="clr-namespace:RpcScandinavia.Core.MVVM.AvaloniaUI"
	x:Class="RpcScandinavia.Core.MVVM.AvaloniaUI.IconButton"
	x:Name="self"
>
	<ContentControl.Styles>
		<Style Selector="ui|IconButton">
			<Setter Property="Template">
				<ControlTemplate>
					<Button
						x:Name="PART_Button"
					>
						<Button.Content>
							<StackPanel Orientation="Horizontal">
								<Image
									Height="18"
									Margin="0, 0, 5, 0"
									HorizontalAlignment="Center"
									VerticalAlignment="Center"
									IsVisible="true"
									Source="{Binding Path=Image, ElementName=self }"
								/>
								<ContentPresenter
									Content="{TemplateBinding Content}"
								/>
							</StackPanel>
						</Button.Content>
					</Button>
				</ControlTemplate>
			</Setter>
		</Style>
	</ContentControl.Styles>
</ContentControl>
```

Code behind in "IconButton.axaml.cs"

```
using System;
using System.Collections.Generic;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Controls.Metadata;
using Avalonia.Controls.Primitives;
using Avalonia.Controls.Templates;
using Avalonia.Interactivity;
using Avalonia.Svg.Skia;
namespace RpcScandinavia.Core.MVVM.AvaloniaUI;

/// <summary>
/// Panel with icon to the left and the content to the right.
/// </summary>
[TemplatePart("PART_Button", typeof(Button))]
public partial class IconButton : ContentControl {
	private readonly HashSet<String> explicitlySetProperties;
	private Button Button { get; set; }

	public IconButton() {
		this.explicitlySetProperties = new HashSet<String>();
		this.InitializeComponent(true);
		this.Image = null;
		this.ImageIsVisible = false;
		// Test icon.
		this.Image = RpcScandinavia.Core.Resources.FluentEmoji.HouseColor.ToAvaloniaSvgImage();
	} // IconButton

	protected override void OnPropertyChanged(AvaloniaPropertyChangedEventArgs change) {
		base.OnPropertyChanged(change);
		this.explicitlySetProperties.Add(change.Property.Name);
	} // OnPropertyChanged

	protected override void OnApplyTemplate(TemplateAppliedEventArgs e) {
		base.OnApplyTemplate(e);

	    this.Button = e.NameScope.Find<Button>("PART_Button");
	    if (this.Button != null) {
			// Copy default values from internal Button to IconButton.
			if (this.explicitlySetProperties.Contains(nameof(this.Background)) == false) this.Background = this.Button.Background;
	        if (this.explicitlySetProperties.Contains(nameof(this.Foreground)) == false) this.Foreground = this.Button.Foreground;
	        if (this.explicitlySetProperties.Contains(nameof(this.BorderBrush)) == false) this.BorderBrush = this.Button.BorderBrush;
	        if (this.explicitlySetProperties.Contains(nameof(this.BorderThickness)) == false) this.BorderThickness = this.Button.BorderThickness;
	        //if (this.explicitlySetProperties.Contains(nameof(this.Command)) == false) this.Command = this.Button.Command;
	        //if (this.explicitlySetProperties.Contains(nameof(this.CommandParameter)) == false) this.CommandParameter = this.Button.CommandParameter;
	        if (this.explicitlySetProperties.Contains(nameof(this.IsEnabled)) == false) this.IsEnabled = this.Button.IsEnabled;
	        if (this.explicitlySetProperties.Contains(nameof(this.FontSize)) == false) this.FontSize = this.Button.FontSize;
	        if (this.explicitlySetProperties.Contains(nameof(this.FontWeight)) == false) this.FontWeight = this.Button.FontWeight;
	        if (this.explicitlySetProperties.Contains(nameof(this.FontStyle)) == false) this.FontStyle = this.Button.FontStyle;
	        if (this.explicitlySetProperties.Contains(nameof(this.FontFamily)) == false) this.FontFamily = this.Button.FontFamily;
	        if (this.explicitlySetProperties.Contains(nameof(this.Padding)) == false) this.Padding = this.Button.Padding;
	        if (this.explicitlySetProperties.Contains(nameof(this.HorizontalAlignment)) == false) this.HorizontalAlignment = this.Button.HorizontalAlignment;
	        if (this.explicitlySetProperties.Contains(nameof(this.VerticalAlignment)) == false) this.VerticalAlignment = this.Button.VerticalAlignment;
	        if (this.explicitlySetProperties.Contains(nameof(this.Width)) == false) this.Width = this.Button.Width;
	        if (this.explicitlySetProperties.Contains(nameof(this.Height)) == false) this.Height = this.Button.Height;
	        if (this.explicitlySetProperties.Contains(nameof(this.MinWidth)) == false) this.MinWidth = this.Button.MinWidth;
	        if (this.explicitlySetProperties.Contains(nameof(this.MinHeight)) == false) this.MinHeight = this.Button.MinHeight;
	        if (this.explicitlySetProperties.Contains(nameof(this.MaxWidth)) == false) this.MaxWidth = this.Button.MaxWidth;
	        if (this.explicitlySetProperties.Contains(nameof(this.MaxHeight)) == false) this.MaxHeight = this.Button.MaxHeight;
	        if (this.explicitlySetProperties.Contains(nameof(this.Margin)) == false) this.Margin = this.Button.Margin;
	        if (this.explicitlySetProperties.Contains(nameof(this.Opacity)) == false) this.Opacity = this.Button.Opacity;
	        //if (this.explicitlySetProperties.Contains(nameof(this.IsDefault)) == false) this.IsDefault = this.Button.IsDefault;
	        //if (this.explicitlySetProperties.Contains(nameof(this.IsCancel)) == false) this.IsCancel = this.Button.IsCancel;
	        //if (this.explicitlySetProperties.Contains(nameof(this.ClickMode)) == false) this.ClickMode = this.Button.ClickMode;

	        // Bind properties from IconButton to internal Button.
	        this.Button.Bind(Button.BackgroundProperty, this.GetObservable(IconButton.BackgroundProperty));
	        this.Button.Bind(Button.ForegroundProperty, this.GetObservable(IconButton.ForegroundProperty));
	        this.Button.Bind(Button.BorderBrushProperty, this.GetObservable(IconButton.BorderBrushProperty));
	        this.Button.Bind(Button.BorderThicknessProperty, this.GetObservable(IconButton.BorderThicknessProperty));
	        //this.Button.Bind(Button.CommandProperty, this.GetObservable(IconButton.CommandProperty));
	        //this.Button.Bind(Button.CommandParameterProperty, this.GetObservable(IconButton.CommandParameterProperty));
	        this.Button.Bind(Button.IsEnabledProperty, this.GetObservable(IconButton.IsEnabledProperty));
	        this.Button.Bind(Button.FontSizeProperty, this.GetObservable(IconButton.FontSizeProperty));
	        this.Button.Bind(Button.FontWeightProperty, this.GetObservable(IconButton.FontWeightProperty));
	        this.Button.Bind(Button.FontStyleProperty, this.GetObservable(IconButton.FontStyleProperty));
	        this.Button.Bind(Button.FontFamilyProperty, this.GetObservable(IconButton.FontFamilyProperty));
	        this.Button.Bind(Button.PaddingProperty, this.GetObservable(IconButton.PaddingProperty));
	        this.Button.Bind(Button.HorizontalAlignmentProperty, this.GetObservable(IconButton.HorizontalAlignmentProperty));
	        this.Button.Bind(Button.VerticalAlignmentProperty, this.GetObservable(IconButton.VerticalAlignmentProperty));
	        this.Button.Bind(Button.WidthProperty, this.GetObservable(IconButton.WidthProperty));
	        this.Button.Bind(Button.HeightProperty, this.GetObservable(IconButton.HeightProperty));
	        this.Button.Bind(Button.MinWidthProperty, this.GetObservable(IconButton.MinWidthProperty));
	        this.Button.Bind(Button.MinHeightProperty, this.GetObservable(IconButton.MinHeightProperty));
	        this.Button.Bind(Button.MaxWidthProperty, this.GetObservable(IconButton.MaxWidthProperty));
	        this.Button.Bind(Button.MaxHeightProperty, this.GetObservable(IconButton.MaxHeightProperty));
	        this.Button.Bind(Button.MarginProperty, this.GetObservable(IconButton.MarginProperty));
	        this.Button.Bind(Button.OpacityProperty, this.GetObservable(IconButton.OpacityProperty));
	        //this.Button.Bind(Button.IsDefaultProperty, this.GetObservable(IconButton.IsDefaultProperty));
	        //this.Button.Bind(Button.IsCancelProperty, this.GetObservable(IconButton.IsCancelProperty));
	        //this.Button.Bind(Button.ClickModeProperty, this.GetObservable(IconButton.ClickModeProperty));

	        // Add event handlers.
	        this.Button.Click += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.ClickEvent, this));
	        this.Button.PointerPressed += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.PointerPressedEvent, this));
	        this.Button.PointerReleased += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.PointerReleasedEvent, this));
	        //this.Button.PointerEnter += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.PointerEnterEvent, this));
	        //this.Button.PointerLeave += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.PointerLeaveEvent, this));
	        //this.Button.PointerMoved += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.PointerMovedEvent, this));
	        this.Button.PointerWheelChanged += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.PointerWheelChangedEvent, this));
	        this.Button.PointerCaptureLost += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.PointerCaptureLostEvent, this));
	        this.Button.KeyDown += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.KeyDownEvent, this));
	        this.Button.KeyUp += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.KeyUpEvent, this));
	        this.Button.GotFocus += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.GotFocusEvent, this));
	        this.Button.LostFocus += (sender, args) => this.RaiseEvent(new RoutedEventArgs(Button.LostFocusEvent, this));
	    }
	} // OnApplyTemplate

	//------------------------------------------------------------------------------------------------------------------
	// Attribute properties.
	//------------------------------------------------------------------------------------------------------------------
	public static readonly StyledProperty<SvgImage> ImageProperty = AvaloniaProperty.Register<IconButton, SvgImage>(nameof(IconButton.ImageProperty));

	/// <summary>
	/// The SVG image that is shown in the image part of the button.
	/// </summary>
	public SvgImage Image {
		get => GetValue(IconButton.ImageProperty);
		set => SetValue(IconButton.ImageProperty, value);
	} // Image

	public static readonly StyledProperty<Boolean> ImageIsVisibleProperty = AvaloniaProperty.Register<IconButton, Boolean>(nameof(IconButton.ImageIsVisibleProperty));

	/// <summary>
	/// Whether the image part of the button is visible.
	/// </summary>
	public Boolean ImageIsVisible {
		get => GetValue(IconButton.ImageIsVisibleProperty);
		set => SetValue(IconButton.ImageIsVisibleProperty, value);
	} // ImageIsVisible

} // IconButton
```
